src/libraries/System.Private.Runtime.InteropServices.JavaScript/tests/System/Runtime/InteropServices/JavaScript/JavaScriptTests.cs

```diff
@@ -115,5 +118,106 @@ public static void FunctionMath()
         }
+
+        [Fact]
+        public static async Task BagIterator()
+        {
+            await Task.Delay(1);
+
+            var bagFn = new Function(@"
+                    var same = {
+                        x:1
+                    };
+                    return Object.entries({
+                        a:1,
+                        b:'two',
+                        c:{fold:{}},
+                        d:same,
+                        e:same,
+                        f:same
+                    });
+                ");
+
+            for (int attempt = 0; attempt < 100_000; attempt++)
+            {
+                try
+                {
+                    using var bag = (JSObject)bagFn.Call(null);
+                    using var entriesIterator = (JSObject)bag.Invoke("entries");
+
+                    var cnt = entriesIterator.ToEnumerable().Count();
+                    Assert.Equal(6, cnt);
+
+                    // fill GC helps to repro
+                    var x = new byte[100 + attempt / 100];
+                }
+                catch (Exception ex)
+                {
+                    throw new Exception(ex.Message + " At attempt=" + attempt, ex);
+                }
+            }
+        }
+
+        [Fact]
+        public static async Task Iterator()
+        {
+            await Task.Delay(1);
+
+            var makeRangeIterator = new Function("start", "end", "step", @"
+                    let nextIndex = start;
+                    let iterationCount = 0;
+
+                    const rangeIterator = {
+                       next: function() {
+                           let result;
+                           if (nextIndex < end) {
+                               result = { value: {}, done: false }
+                               nextIndex += step;
+                               iterationCount++;
+                               return result;
+                           }
+                           return { value: {}, done: true }
+                       }
+                    };
+                    return rangeIterator;
+                ");
+
+            for (int attempt = 0; attempt < 100_000; attempt++)
+            {
+                try
+                {
+                    using (var entriesIterator = (JSObject)makeRangeIterator.Call(null, 0, 500))
+                    {
+                        var cnt = entriesIterator.ToEnumerable().Count();
+                    }
+                }
+                catch (Exception ex)
+                {
+                    throw new Exception(ex.Message + " At attempt=" + attempt, ex);
+                }
+            }
+        }
+
+        public static IEnumerable<object> ToEnumerable(this JSObject iterrator)
+        {
+            JSObject nextResult = null;
+            try
+            {
+                nextResult = (JSObject)iterrator.Invoke("next");
+                var done = (bool)nextResult.GetObjectProperty("done");
```

Comment:
Can nextResult be null? I think it can be, so we should check.

Reply:
It could be only by bug. But the iterator API doesn't allow it.
https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Iteration_protocols#the_iterator_protocol
```
The next() method must always return an object with appropriate properties including done and value. If a non-object value gets returned (such as false or undefined), a TypeError will be thrown.
```